src/dmd/must_use.d

```diff
@@ -0,0 +1,203 @@
+module dmd.must_use;
+
+import dmd.dscope;
+import dmd.dsymbol;
+import dmd.expression;
+
+/**
+ * Check whether discarding an expression would violate the requirements of
+ * @mustUse. If so, emit an error.
+ *
+ * Params:
+ *   e = the expression to check
+ *   sc = scope in which `e` was semantically analyzed
+ *
+ * Returns: true on error, false on success.
+ */
+bool checkMustUse(Expression e, Scope* sc)
+{
+    import dmd.id : Id;
+
+    if (!isAssignment(e) && !isIncrementOrDecrement(e))
+    {
+        assert(e.type);
+        auto sym = e.type.toDsymbol(sc);
+        auto sd = sym ? sym.isStructDeclaration() : null;
+        // isStructDeclaration returns non-null for both structs and unions
+        if (sd && hasMustUseAttribute(sd, sc))
+        {
+            e.error("ignored value of `@%s` type `%s`; prepend a `cast(void)` if intentional",
+                Id.udaMustUse.toChars(), e.type.toChars());
```

Comment:
Any specific reason to not use the fully-qualified name (`toPrettyChars`)? Might be sensible given that this feature targets external API's using e.g. error codes.